### DISCUSSION TITLE
RefreshDatabase running migrate:fresh once per test suite

### DISCUSSION THREAD

**Discussion by sebastianluy** [laravel/framework#41750] [Q&A]
Hello all,

I'm trying to understand the RefreshDatabase trait and how it uses migrations. I read in the documentation that:

"The Illuminate\Foundation\Testing\RefreshDatabase trait does not migrate your database if your schema is up to date."
https://laravel.com/docs/9.x/database-testing#resetting-the-database-after-each-test

Based off what I'm seeing in my project and looking at the source code, it looks like the database is migrated with `artisan migrate:fresh`, although only once per run of the test suite. https://github.com/laravel/framework/blob/b9203fca96960ef9cd8860cb4ec99d1279353a8d/src/Illuminate/Foundation/Testing/RefreshDatabase.php#L70

The migration is conditional on the `RefreshDatabaseState::$migrated` flag, but it doesn't look like the application ever sets that value before running the migrations the first time.

Is this the intended behavior? If so, the documentation seems kind of misleading here. Maybe it should read:

"The Illuminate\Foundation\Testing\RefreshDatabase trait will migrate (migrate:fresh) your database only once per run of the test suite."

**Comment by beybol** [comment]
I am facing the same problem. My schema is up to date and migrate:fresh is executed every time I run tests. It results in database wipe on every run.

**Comment by bmellink** [comment]
I have the same issue. Anyone aware this has been resolved? (laravel 9). I found a package at https://github.com/PlannrCrm/laravel-fast-refresh-database who wrote an alternative for the refreshdatabase trait that is 'smart'

**Comment by cedrictwillie** [comment]
Being your schema should not be modified at run time, the ideal setup I am looking for is:

1. Run migrations "only once" before any test case is run.
2. Run "initial data" seeders "only once" before any test case is run.
3. Then run each test, beginning a transaction to rollback any data modifications made after the initial seeding.

I am finding this difficult to implement with a multi database instance architecture.
Ideally, I never want to re-run migrations after the first and only run on all databases. I find this to be a huge performance hit for my setup.
